components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java

```diff
@@ -0,0 +1,123 @@
+   * {@link Content#getAttributes()}) from passed in {@link Asset} and format required hashes.
+   */
+  public static void extractFromAsset(final Asset asset,
+                                      final Iterable<HashAlgorithm> hashAlgorithms,
+                                      final AttributesMap contentAttributes)
+  {
+    checkNotNull(asset);
+    checkNotNull(hashAlgorithms);
+    final NestedAttributesMap assetAttributes = asset.attributes().child(P_CONTENT);
+    final DateTime lastModified = toDateTime(assetAttributes.get(P_LAST_MODIFIED, Date.class));
+    final String etag = assetAttributes.get(P_ETAG, String.class);
+
+    final NestedAttributesMap checksumAttributes = asset.attributes().child(StorageFacet.P_CHECKSUM);
+    final Map<HashAlgorithm, HashCode> hashCodes = Maps.newHashMap();
+    for (HashAlgorithm algorithm : hashAlgorithms) {
+      final HashCode hashCode = HashCode.fromString(checksumAttributes.require(algorithm.name(), String.class));
+      hashCodes.put(algorithm, hashCode);
+    }
+
+    contentAttributes.set(Content.CONTENT_LAST_MODIFIED, lastModified);
+    contentAttributes.set(Content.CONTENT_ETAG, etag);
+    contentAttributes.set(Content.CONTENT_HASH_CODES_MAP, hashCodes);
+    contentAttributes.set(CacheInfo.class, CacheInfoMarshaller.extractFromAsset(asset));
+  }
+
+  /**
+   * Applies non-format specific content attributes onto passed in {@link Asset} from passed in {@link AttributesMap}
+   * (usually originating from {@link Content#getAttributes()}).
+   */
+  public static void applyToAsset(final Asset asset, final AttributesMap contentAttributes) {
+    checkNotNull(asset);
+    checkNotNull(contentAttributes);
+    final NestedAttributesMap assetAttributes = asset.attributes().child(P_CONTENT);
+    assetAttributes.set(P_LAST_MODIFIED, toDate(contentAttributes.get(Content.CONTENT_LAST_MODIFIED, DateTime.class)));
+    assetAttributes.set(P_ETAG, contentAttributes.get(Content.CONTENT_ETAG, String.class));
+    final CacheInfo cacheInfo = contentAttributes.get(CacheInfo.class);
+    if (cacheInfo != null) {
+      CacheInfoMarshaller.applyToAsset(asset, cacheInfo);
+    }
+  }
+
+  /**
+   * Applies non-format specific content attributes onto passed in {@link Asset} from passed in {@link AttributesMap}
+   * (usually originating from {@link Content#getAttributes()}). The {@code contentAttributes} input may be null too,
+   * in which case this method will create empty content attributes. If the content attributes (passed in or new) does
+   * not contains {@link Content#CONTENT_LAST_MODIFIED} mapping, mapping with "now" timestamp will be added.
+   *
+   * @see #applyToAsset(Asset, AttributesMap)
+   */
+  public static void applyToAssetMaintainLastModified(final Asset asset,
+                                                      @Nullable final AttributesMap contentAttributes)
```

Comment:
Method name suggests this is trying to do too much - how about a method `maintainLastModified` which takes an asset and attributes map and just maintains the last-modified time, which clients can then apply using `applyToAsset`.